doc/agaconv.1
.\" Automatically generated by Pandoc 2.9.2.1
.\"
.TH "AGACONV" "1" "Jul 2024" "agaconv" "Version 1.1"
.hy
.SH NAME
.PP
agaconv - converts videos into CDXL format for AGA Commodore-Amiga computers
.SH SYNOPSIS
.PP
agaconv INPUTVIDEO CDXLVIDEO [\f[I]OPTION\f[R]]
.SH DESCRIPTION
.PP
\f[B]AGAConv\f[R] is a CDXL video converter.
The input video can be in any format that can be processed by ffmpeg.
\f[B]AGAConv\f[R] uses ffmpeg to extract video and audio data and encodes the
extracted frames into the CDXL video format.
\f[B]AGAConv\f[R] can generate standard and custom CDXL files.
By default \f[B]AGAConv\f[R] generates standard CDXL files.
It also ensure that the CDXL chunks are 32bit aligned which can improve video
frame rate by up to 50% on some Amiga I/O controllers.
If padding is selected then custom CDXL files are generated.
.PP
The CDXL format was created by Commodore primarily for the Commodore CDTV,
released 1991, to permit playback of video from CD-ROM in the early 1990s.
The CDXL format is a simple streaming format, consisting of linear concatenated
chunks (packets), each with an uncompressed frame and associated audio data.
In addition to the original 12-bit colors, \f[B]AGAConv\f[R] also supports
24-bit colors, as supported by Commodore Amiga AGA systems released in 1992+.
\f[B]AGAConv\f[R] also ensures 32-bit alignment of CDXL chunks, which can make a
difference in video frame rate of up to 50%.
For standard CDXL files this is achieved by adjusting the frequency.
See option --frequency for more details.
.PP
Standard 12-bit colors CDXL video files can be played with any CDXL player.
AGABlaster is a CDXL player that also supports 24-bit colors CDXL videos and
utilizes the AGA 64-bit graphics access for best frame rate on Commodore-Amiga
AGA systems.
.PP
The tool ffmpeg is automatically installed with the agaconv PPA installer.
\f[B]AGAConv\f[R] can be used for AGA8 to AGA2 and OCS5 to OCS2 conversions
without installing any additional tools.
From those screen modes, AGA8 gives the best quality on AGA systems.
.PP
For converting videos into HAM6, HAM8, or EHB CDXL videos, the additional tool
ham_convert must be installed.
See option --hc-path for how to set the path to ham_convert in the
\f[B]AGAConv\f[R] config file and Section `SEE ALSO' for web links.
Once configured, \f[B]AGAConv\f[R] integrates ham_convert into the conversion
pipeline.
.PP
\f[B]AGAConv\f[R] also supports configuration files.
A configuration file can be used to set all options listed in section OPTIONS
for each video conversion parameter.
A config file can be edited with a text editor.
Entries in a config file have the same name as its corresponding command line
option, with the only difference that a dash in the command line option is
written is an underscore in the config file.
For example, --color-mode=aga8 is listed as color_mode = aga8 in a config file.
Only existing entries in a config file set a value, for non-existing entries the
(original) default value is used.
Command line options always override config file options.
Entries in a config file without a value are handled the same way as
non-existing entries.
.PP
\f[B]AGAConv\f[R] can store a default config file in a default location.
The entries in the default config file can be changed through the command line
with the --change-config option or the file can be edited with a text editor.
The default config file is automatically generated when the option
--change-config is used for the first time.
.PP
\f[B]AGAConv\f[R] also supports user config files.
When a user config file is loaded it is used instead of the default
configuration.
See the various config file options and Section Examples for more details.
.PP
All conversion options have default values.
A video can be converted without any additional options.
.SH OPTIONS
.PP
agaconv takes the following options:
.SS Basic Options
.TP
--format STRING
CDXL format where STRING = ctm-opt|std-opt|std-fixed (default: std-opt).
This option selects specific properties of CDXL chunks.
All three options ensure 32-bit alignment of all chunks.
This improves display speed.
.RS
.PP
Format option `std-opt' is the default option and selects standard CDXL with
optimized chunk size.
CDXL allows to use chunks of different size in each frame, this option takes
advantage of this CDXL feature and minimizes file size by not including empty
planes and also reducing the color map size if possible.
.PP
Format option `std-fixed' is a compatibility option for older CDXL players that
require that each frame has the exact same size.
\f[B]AGAConv\f[R] adjusts the provided frequency if necessary, to generate a
video where each frame has the exact same size.
In addition, it also ensures that all chunks in the CDXL file are 32-bit
aligned, which can improve I/O speed of up to 50% with certain I/O controllers.
Therefore the length of each audio data and graphics data chunk must be
divisible by 4, to ensure that the next chunk starts at a 32-bit aligned offset
in the CDXL file.
Bit depth is chosen according to the color mode, agaX selects 24 bit mode.
A 12-bit mode can be enforced for agaX modes with the extra option
--force-color-depth=12.
.PP
The format option `ctm-opt' aims to generate the smallest file size possible and
also maintains the exact specified frequency (no adjustment), but also maintains
32-bit alignment of all chunks by introducing padding where necessary (1 to 3
bytes).
.PP
AGABlaster can play all three CDXL variants with any combination of other
options.
.RE
.TP
--color-mode STRING
Color mode where STRING = aga8|..|aga2|ham8|ham6|ocs5|..|ocs2|ehb.
The letters specify the screen mode, the digit determines the number of
bitplanes.
The higher the digit, the better is the quality and the more graphics data is
stored in the CDXL video file.
`agaX' selects 24-bit colors, OCS (Original Chip Set) selects 12-bit colors.
The `aga' and `ocs' modes work out-of-the-box after installation of agaconv.
The three modes ham6, ham8 and ehb require the additional tool ham_convert to be
installed (also see option --hc-path).
.TP
--fps NUMBER
Frames per second.
The frames per second options defines how much graphics data is stored per
second in the CDXL video file.
The higher the FPS, the more data is stored, and the better is the quality due
to a higher number of frames per second.
However, if the amount of data is beyond the I/O bandwidth of the respective
storage device from which the CDXL video file is loaded, the frame rate will be
limited by the bandwidth.
The CDXL player AGABlaster will skip the additional frames on disk and not load
them into memory, effectively adapting the frame rate (and quality) to the I/O
bandwidth.
.TP
--width NUMBER
Width of video in pixels.
The Amiga AGA graphics hardware supports three different resolutions, Lores,
Hires, Superhires with 2 to 8 bitplanes.
Maximum standard width for Lores is 320, for Hires 640, and for Superhires 1280.
For the three specific widths of 384, 704, and 1344, overscan videos are
generated (overscan is maximum standard width plus up to 64 pixels).
Minimum video width is 128 and the width must be a multiple of 16.
Note that AGABlaster gives the best frame rate if the width of a video is a
multiple of 64, because this enables the 64 bit graphics access of the Amiga AGA
architecture.
.TP
--height auto|NUMBER
With `auto' the height of the video is derived from the input video\[cq]s aspect
ratio.
Otherwise the absolute value NUMBER is used.
.TP
--frequency NUMBER
Set audio frequency.
The Audio frequency determines how many bytes per second are used for the audio
data of the video.
The higher the frequency, the better is the audio quality.
Maximum value with the original Amiga audio hardware is 28836.
.TP
--audio-mode mono|stereo
Audio mode.
The two modes mono and stereo can be selected.
Default mode is stereo.
In stereo mode twice the amount of audio data is used in comparison to mono.
.TP
--adjust-aspect FLOAT|hdstretched
This option allows to adjust the aspect ratio of the video.
The default value of 1.0 keeps the ratio of width to height unmodified.
A value greater than 1.0 stretches the video in height.
.RS
.PP
With `hdstretched' the predefined value of 1.35 is used.
This value should work on most modern HD screens when using `fill' or
`stretched' mode to display the Amiga\[cq]s 4:3 screen on an HD screen.
For other monitor configurations a floating point value FLOAT can be used.
.RE
.TP
--hc-path PATH
Absolute file path to the directory of the conversion tool ham_convert.
This setting is only necessary for the screen modes HAM6, HAM8, EHB.
All other modes do not require ham_convert and have a 10-15 times faster
conversion time.
However, for enjoying the HAM modes ham_convert is the only option.
For this purpose the path to ham_convert needs to be added to the configuration
(or provided on the command line).
.RS
.PP
For example, if ham_convert_X_Y_Z.zip was unpacked in the directory /home/john
then the default path to ham_convert can be set with:
.PP
agaconv \[en]change-config \[en]hc-path=/home/john/ham_convert
.PP
This stores the path in the default config file and agaconv will use this path
now with every invocation of agaconv to find ham_convert\[cq]s Java jar file
inside the directory /home/john/ham_convert.
When this path is set, agaconv checks if ham_convert/ham_convert_X.Y.Z.jar
exists.
If it does not exist, it immediately issues an error message.
Hence, when above command succeeds, one can be sure that the provided path is
indeed correct.
One can also specify the full path, including the name of the ham_convert\[cq]s
jar file itself,
e.g.\ \[en]hc-path=\[lq]/home/john/ham_convert/ham_convert_1.9.0.jar\[rq]
.PP
However, to simplify updates of ham_convert, agaconv does support to search for
the Java jar file inside the ham_convert directory, and therefore it\[cq]s best
to only specify the path with the directory name `ham_convert' alone.
.PP
When updating ham_convert, one should first delete the ham_convert directory and
then unzip the ham_convert_X_Y_Z.zip file.
If one accidentally unzips a new version into the directory of an old version,
\f[B]AGAConv\f[R] will detect that there is more than one jar file, and will
issue an error message to make sure one does not accidentally use an old version
of ham_convert.
.RE
.TP
--cdxl-info FILE
Show all info of frame 1 of given CDXL video FILE.
This option is useful for checking the values of converted CDXL videos.
Example: agaconv \[en]cdxl-info video.cdxl
.TP
--cdxl-info-all FILE
Show info of all frames in given CDXL video.
.TP
--verbose NUMBER
Select how verbose the output is during conversion.
The value 0 means that no information is printed during conversion, except error
messages.
Verbose level 1 prints about 5-10 lines for one converted video.
Verbose level 2 and 3 produce additional output for each converted frame.
.TP
--version
print program version and copyright.
.TP
--help
show basic command line options with default values.
If a config file is used then the default values are read from the config file
and shown with this command.
.TP
--help-advanced
show advanced command line options with default values.
If a config file is used then the advanced default values are read from the
config file and shown with this command.
.SS Advanced Options
.TP
--black-and-white
Convert video to black-and-white colors.
This mode uses ffmpeg\[cq]s `gray' option.
.TP
--black-background
Reserve black background color.
This is only relevant on OCS systems, on which one may see a flashing background
if the background color register $0180 is not black and changing color
throughout the video.
On AGA systems CDXL players such as AGABlaster use the special AGA mode with a
blank background, and therefore the background color can be used as well without
producing a flashing background.
.TP
--dither STRING
Sets ffmpeg dithering mode when rescaling video, where
STRING=floyd_steinberg|bayer:bayer_scale=X|sierra2.
This option allows to use any dithering mode that ffmpeg supports.
It appears that floyd_steinberg works best for most videos.
However, since this string is simply passed through to ffmpeg one can experiment
also with other ffmpeg dithering modes.
.TP
--screen-mode STRING
screen (resolution) mode, where STRING =
auto|unspecified|lores|hires|superhires.
The default setting is `auto', which means that the screen resolution is derived
from the provided width parameter.
For example if the provided width it less than or equal to 320 lores resolution
is selected.
If a value higher than 320 and less or equal 640 is selected then hires modes is
selected.
Similar with `superhires'.
The setting `unspecified' should be used when converting CDXL videos for
non-Amiga systems, to ensure that the video height is not rescaled as it is
required for some Amiga native screen modes.
.TP
--force-color-depth auto|12|24
The default setting is `auto'.
The color depth is set automatically when the option \[en]color-mode is used or
by default, such that the best quality is achieved.
Therefore this option can only be used to force reduced quality to 12bit, or to
use 24bit encoding when it does not improve quality because the selected color
mode only uses 12bit color range.
Some Amiga games did indeed use AGA8 videos with 8 planes in 12bit color mode.
This setting allows to reproduce those settings.
However, in general this setting gives no improvement in quality and only exists
for experimentation.
.RS
.PP
Note: --color-mode=ehb --force-color-mode=24 would create a non-existing Amiga
display mode and is rejected.
All other combinations are supported.
.RE
.TP
--install-config
Install the default config file in the respective OS specific location.
On Linux this location is \[ti]/.config/agaconv/default.config.
The option --change-config automatically installs a config file if it was not
installed yet.
.TP
--uninstall-config
Uninstall the default config file.
With this option \f[B]AGAConv\f[R] removes the default config file as well as
the directory agaconv from the ./config directory.
Note this operation is only successful if no other file is in the
\&./config/agaconv directory, e.g.\ after editing the default config file there
might be some backup files in this directory.
The directory is only removed if it is empty after removing the default config
file.
.TP
--reset-config
Reset the default config file to original values.
This reestablishes the original default values in the default config file, but
keeps the hc_path entry, if it was set.
All other values are reset.
Keeping the hc_path is a convenience feature, because this is the only value
that must be set manually after installation if one wishes to use ham_convert
for HAM conversions.
Therefore this option also reads the existing default config file before
regenerating a new default config file.
.TP
--change-config
Change default configuration settings.
Any other option can be added on the command line and the respective value is
then set as default value in the default configuration.
For example, it can be used with other options, e.g.\ \[en]hc-path=PATH, to set
the path for the ham_convert tool.
It can also be used to set the default width to 640.
.RS
.PP
Example: agaconv --change-config --width=640
.PP
This sets the default width of a converted video to a width of 640.
Option --help shows with each option also its default values and can therefore
be used to check the effect of --change-config to the default values.
.RE
.TP
--load-config FILE
Load a user configuration file with name FILE instead of the default
configuration file.
.TP
--save-config FILE
Save the current configuration as FILE.
This basically saves the current default configuration in a separate file.
Additional options can be specified on the command line.
.TP
--tmp-dir-prefix DIRNAME
Sets the prefix of the temporary directory name as DIRNAME.
The default name is \[lq]tmp-agaconv\[rq] and it is recommended to keep this
setting unmodified.
.RS
.PP
During conversion \f[B]AGAConv\f[R] generates by default a local directory with
this name as prefix from where the command `agaconv' was invoked.
It extends this prefix with the name of the video and some more parameters and
the process ID, to ensure the name of the directory is unique.
This allows to run multiple instances of \f[B]AGAConv\f[R] in parallel without
writing into the same directory.
.PP
DIRNAME can also be set to an absolute path, such as
--tmp-dir-prefix=/home/user/tmp-agaconv --change-config
.PP
The temporary directory is removed after every conversion.
It is only removed if it is indeed empty after all generated files have been
removed.
\f[B]AGAConv\f[R] does not remove any other files than those that ffmpeg or
ham_convert can generate for a given video file and some extra files that
\f[B]AGAConv\f[R] itself generates during conversion.
If an error occurs or the process is killed during conversion, then the
temporary directory is not removed.
.PP
With option --verbose=2 \f[B]AGAConv\f[R] prints additional information about
the removed files at the end of a conversion.
.RE
.TP
--keep-tmp-dir
Keep the temporary directory.
The temporary directory is removed by default after each conversion.
This option is only relevant if one wants to inspect the generated PNG or IFF
files that are generated for each frame.
.TP
--hc-ham-quality NUMBER
This is a ham_convert HAM quality option for setting the quality level in the
HAM generation.
Default is 1 and the range for HAM8 is 0..3.
Values greater or equal 2 are significantly slower and require a lot of patience
for longer videos.
Therefore the default is 1.
.TP
--hc-dither STRING
ham_convert dither mode where STRING=auto|none|fs|bayer8x8.
The default value is `fs' (Floyd-Steinberg algorithm, as with ffmpeg).
If `none' is selected then no dither is applied.
With `auto' the default ham_convert mode is used (which currently is `none').
.TP
--hc-propagation NUMBER
ham_convert error propagation factor, requires hc_dither = fs, default: `auto',
value range is 0..100.
See ham_convert documentation for more details.
.TP
--hc-diversity NUMBER
ham_convert diversity X=0-6 for ehb, X=0-9 for other modes, not supported in
ham8.
Default is `auto', value range is 0..9.
See ham_convert documentation for more details.
.TP
--hc-quant STRING
ham_convert quantization algorithm where STRING=wu|neuquant.
Default setting is `auto'.
See ham_convert documentation for more details.
.TP
--iff-info FILE
Show IFF file info for a given IFF FILE.
This can be used to inspect IFF files generated with ham_convert.
To keep the IFF files at the end of a conversion, the option --keep-tmp-dir can
be used.
.TP
--in-file FILE
Set the input file name.
This option is available for tool generated config files.
.TP
--out-file FILE
Set the output file name.
This option is available for tool generated config files.
.SH ENVIRONMENT
.PP
AGAConv uses the following environment variables
.TP
HOME
This environment variable is used to determine the user\[cq]s home directory and
store a default config file as $HOME/.config/agaconv/default.config
.RS
.PP
The default config file is generated when the options --install-config or
--change-config are used.
It can be removed with --uninstall-config
.RE
.SH EXIT STATUS
.TP
Exit status 0
If no error occurs.
.TP
Exit status 1
If any error occurs due to wrong command line arguments, errors in the default
config file, or during conversion, an error message is printed and the program
exits with return code 1.
If any invoked external tool fails, AGAConv also returns an exit status of 1 and
reports the invoked tool\[cq]s status return code in the error message.
.RS
.PP
Every error is reported with an identifying error number and an error message on
stderr.
.RE
.SH EXAMPLES
.TP
agaconv video.mp4 video.cdxl
Converts the mpeg video into a CDXL video.
By default the CDXL video is encoded with a width of 320, lores resolution,
24-bit colors, AGA8 with 8 bitplanes (256 colors per frame), 24 FPS, and stereo
audio with 28000 Hz.
.TP
agaconv video.mp4 video.cdxl --fps=25
Encodes the CDXL video with 25 frames per second and default values, using
24-bit colors, AGA8 with 8 bitplanes (256 colors per frame), stereo audio with
28000 Hz,
.TP
agaconv --std-opt video.mp4 video.cdxl --fps=25
Encodes a standard CDXL video with 25 frames per second and optimized frame
size, using 24-bit colors, AGA8 with 8 bitplanes (256 colors per frame), stereo
audio with 28000 Hz,
.TP
agaconv --std-fixed video.mp4 video.cdxl --fps=25
Encodes a standard CDXL video with 25 frames per second and fixed frame size,
using 24-bit colors, AGA8 with 8 bitplanes (256 colors per frame), stereo audio
with 28000 Hz,
.TP
agaconv --ctm-opt video.mp4 video.cdxl --frequency 28112 --fps=25
Encodes a customized CDXL video with 25 frames per second and minimal frame
size, using 24-bit colors, AGA8 with 8 bitplanes (256 colors per frame), stereo
audio with 28112 Hz.
Custom CDXL mode allows to use any frequency, whereas the std modes adjust the
frequency to have 32-bit aligned chunks.
In custom mode data padding is used instead of adjustment.
.TP
agaconv video.mp4 video.cdxl --fps=25 --width=640
Encodes the CDXL video with 25 frames per second and a width of 640 pixels.
\f[B]AGAConv\f[R] derives from the width that the video must be encoded for a
Hires resolution and adjusts the aspect ratio accordingly.
All other parameters have default values, using 24-bit colors, AGA8 with 8
bitplanes (256 colors per frame), and stereo audio with 28000 Hz.
.TP
agaconv video.mp4 video.cdxl --monitor-mode=stretched
Encodes the CDXL video such that it can be played on an HD screen where the
display is stretched (or filled) to full screen.
If the Amiga produces a 4:3 output, this option can be used to stretch the video
(in y-direction) such that when displayed its aspect ratio is correct.
.TP
agaconv --cdxl-info video.cdxl
Shows all entries of the first frame of the CDXL file video.cdxl
.TP
agaconv video.mp4 video.cdxl
Encodes the CDXL video as standard CDXL video with a width of 320.
For standard CDXL videos \f[B]AGAConv\f[R] adjusts the frequency such that the
audio chunk is 32bit aligned.
For AGA modes Standard CDXL is encoded with 24bit colors.
.TP
agaconv video.mp4 video.cdxl --color-mode=aga7
Encodes the CDXL video with 7 bitplanes and 128 colors.
For standard CDXL videos \f[B]AGAConv\f[R] adjusts the frequency such that all
frames have the same size and all data chunks are 32-bit aligned.
By default 24bit colors are used for AGA modes, and 12bit colors for OCS, EHB,
and HAM6 modes.
12bit colors can be enforced also for AGA modes with --force-color-depth=12.
.TP
agaconv video.mp4 video.cdxl --color-mode=ham8
Encodes the CDXL video in HAM8 format.
HAM8 uses 24-bit colors.
This requires the tool ham_convert to be installed and the path to ham_convert
must be set in the \f[B]AGAConv\f[R] config file.
.TP
agaconv video.mp4 video.cdxl --color-mode=ocs5 --fps=20 --frequency=18000 --audio-mode=mono
Encodes the CDXL video with 5 bitplanes (32 colors) with 12-bit colors for an
OCS Amiga system, with 20 frames per second, and a frequency of 18000 Hz in mono
audio mode.
\f[B]AGAConv\f[R] uses dithering by default, which gives best results.
Option --dither allows to experiment with other dithering modes that are
supported by ffmpeg.
.TP
agaconv --width=640 --save-config my-hires-video.config
This sets the width to 640 in the saved configuration file
my-hires-video.config.
When converting a video this can be used with agaconv --load-config
my-hires-video.config inputvideo.mp4 outputvideo.cdxl
.TP
agaconv --width=640 --change-config
Change the default configuration to use 640 as width.
On a classic Amiga this will be displayed as a hires video with width 640 by
\f[B]AGABlaster\f[R].
The option --help will also display as default width value 640 after this change
in the config file.
.TP
agaconv --help
This option shows all basic help options with its respective default values.
The default values are defined by the configuration.
If the default config has been changed by using the option --change-config then
the option --help will list those (changed) default values.
Hence, --help can also be used to check the default values.
.TP
agaconv --load-config my-user.config --change-config
Load the user configuration my-user.config and set it as default configuration.
.TP
agaconv --load-config my-user1.config --save-config my-user2.config
Load my-user1.config and save it as my-user2.config.
This is essentially the same as copying the file, but checks all values to be in
correct ranges.
It also checks that the hc_path (if set) refers to an existing version of
ham_convert.
The --load|save-config options can be combined with the change-config option for
the default configuration.
.SH COPYRIGHT
.PP
Copyright (C) 2019-2024 Markus Schordan.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH SEE ALSO
.IP \[bu] 2
\f[B]AGAConv\f[R] webpage: <https://cutt.ly/AGAConv>
.IP \[bu] 2
\f[B]AGABlaster\f[R] webpage: <https://cutt.ly/AGABlaster>
.SH BUGS
.PP
No known bugs.
.SH AUTHORS
Markus Schordan.


src/agaconv.cpp
/*
    AGAConv - CDXL video converter for Commodore-Amiga computers
    Copyright (C) 2019-2024 Markus Schordan

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <filesystem>
#include <iostream>

#include "CDXLDecode.hpp"
#include "CDXLEncode.hpp"
#include "CommandLineParser.hpp"
#include "Configuration.hpp"
#include "ExternalToolDriver.hpp"
#include "StageAnimEdit.hpp"
#include "StageChunkInfo.hpp"
#include "StageILBMFileInfo.hpp"

using namespace std;
using namespace AGAConv;

void runCDXLEncode(Options& options) {
  CDXLEncode stage;
  std::filesystem::path inFileWithPath=options.getTmpDirName()/("frame"+options.firstFrameNumberToString()+".png");
  stage.setInFileWithPath(inFileWithPath);
  stage.run(options);
}

int main(int argc, char **argv) {
  string version="1.1.0";
  try {
    Configuration config;
    CommandLineParser clp;
    clp.setVersion(version);
    clp.parse(argc,argv,config);
    if(clp.done())
      return 0;
    Options options=config.getOptions();

    ////////////////
    // CDXL files //
    ////////////////
    
    // required
    Chunk::animPaddingFix=options.animPaddingFix;
    Chunk::setDebug(options.debug);
    
    // CDXL encoding
    if(options.writeCdxl && options.cdxlEncode && !options.cdxlInfo && !options.cdxlDecode &&!options.ilbmInfo) {
      // Extraction of png files
      ExternalToolDriver etd;
      etd.checkCommandProcessor();
      if(options.verbose>=1)
        cout<<"Conversion started."<<endl;
      etd.prepareTmpDir(options);
      if(options.verbose>=1)
        cout<<"Converting video file "<<options.inFileName<<endl;
      etd.runFFMPEGExtraction(options);

      // Conversion
      if(options.conversionTool=="ffmpeg") {
        runCDXLEncode(options);
      } else if(options.conversionTool=="ham_convert") {
        etd.runHamConvert(options);
      } else {
        throw AGAConvException(1,"unknown conversion tool "+options.conversionTool);
      }
      etd.finalizeTmpDir(options);
      if(options.verbose>=1)
        cout<<"Conversion finished."<<endl;
      if(options.verbose<=1 && !config.isSupportedOS() && options.conversionTool=="ham_convert") {
        cout<<"WARNING: unsupported OS. verbose level <= 1 not fully supported. Could not suppress output of external tools."<<endl;
      }
      return 0;
    }

    // CDXL file info for one file
    if(options.cdxlInfo) {
      // Read just first frame's header
      // Open file
      auto fileName=options.inFileName;
      fstream inFile;
      inFile.open(fileName, ios::in | ios::binary);
      if(inFile.is_open() == false) {        
        throw AGAConvException(2,"File not found: "+fileName.string());
      }
      CDXLHeader cdxlHeader;
      cdxlHeader.setFile(&inFile);
      cdxlHeader.readChunk();
      cout<<cdxlHeader.toString();
      inFile.close();
      return 0;
    }

    // CDXL file header printing
    if(options.cdxlDecode) {
      CDXLDecode stage;
      stage.run(options);
      return 0;
    }

    // ILBM/IFF file info printing
    if(options.ilbmInfo) {
      StageILBMFileInfo stage;
      stage.run(options);
      return 0;
    }

    ////////////////
    // ANIM files //
    ////////////////

    // Regenerate original file (for testing) or also add sound data if present
    if(options.writeAnim && options.hasOutFile() &&!options.writeCdxl) {
      StageAnimEdit stage;
      stage.run(options);
      return 0;
    }

    // Anim chunk info option
    if(options.chunkInfo) {
      StageChunkInfo stage;
      stage.run(options);
      return 0;
    }

  } catch(AGAConvException& e) {
    cerr<<e.what()<<endl;
    cerr<<"Exiting."<<endl;
    return 1;
  } catch(std::exception& e) {
    cerr<<e.what()<<endl;
    return 1;
  }
  return 0;
}  
